source/common/io/io_uring.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+
+#include "source/common/network/address_impl.h"
+
+namespace Envoy {
+namespace Io {
+
+/**
+ * Callback invoked when iterating over entries in the completion queue.
+ * @param user_data is any data attached to an entry submitted to the submission
+ * queue.
+ * @param result is a return code of submitted system call.
+ */
+using CompletionCb = std::function<void(void* user_data, int32_t result)>;
+
+enum class IoUringResult { Ok, Busy, Failed };
+
+/**
+ * Abstract wrapper around `io_uring`.
```

Comment:
io_uring docs don't state any requirement regarding thread-safety though it probably should. In case of Envoy every IoUring is coupled with one libevent loop (it doesn't run its own) which is rarely posted from threads other than the one running it. So, I assumed that making IoUring thread local and calling its methods only from the thread it lives in simplifies everything. But perhaps it makes more sense to extend Dispatcher to own IoUrings.

Reply:
It may make sense for the dispatcher to own the io_urings eventually instead of providing a getOrCreate factory method.  We can get to that as development continues.